File: server/models/MaintenanceTask.js
const mongoose = require('mongoose');

const maintenanceTaskSchema = new mongoose.Schema({
  taskId: {
    type: String,
    required: true,
    unique: true
  },
  sourceSystem: {
    type: String,
    enum: ['TMS', 'SMMS', 'TDMS'],
    required: true
  },
  department: {
    type: String,
    enum: ['Engineering', 'Traction Distribution', 'Signal & Telecom'],
    required: true
  },
  sectionId: {
    type: String,
    required: true
  },
  sectionName: {
    type: String,
    required: true
  },
  blockId: {
    type: String,
    default: ''
  },
  defectType: {
    type: String,
    required: true
  },
  defectDescription: {
    type: String,
    default: ''
  },
  criticality: {
    type: String,
    enum: ['critical', 'high', 'medium', 'low'],
    default: 'medium'
  },
  reportedDate: {
    type: Date,
    required: true
  },
  dueDate: {
    type: Date,
    required: true
  },
  estimatedDuration: {
    type: Number, // in minutes
    required: true,
    min: 15,
    max: 480
  },
  status: {
    type: String,
    enum: ['pending', 'scheduled', 'approved', 'in_progress', 'completed', 'cancelled'],
    default: 'pending'
  },
  // AI Scoring fields
  criticalityScore: {
    type: Number,
    default: null,
    min: 0,
    max: 1
  },
  scoreBreakdown: {
    safety: { type: Number, default: null },
    overdue: { type: Number, default: null },
    traffic: { type: Number, default: null },
    recurrence: { type: Number, default: null }
  },
  urgencyTier: {
    type: String,
    enum: ['Critical', 'High', 'Medium', 'Low', null],
    default: null
  },
  // Historical data
  recurrenceCount: {
    type: Number,
    default: 0
  },
  lastOccurrence: {
    type: Date,
    default: null
  },
  // Assignment
  assignedTo: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null
  },
  notes: {
    type: String,
    default: ''
  }
}, {
  timestamps: true
});

// Index for common queries
maintenanceTaskSchema.index({ department: 1, status: 1 });
maintenanceTaskSchema.index({ sectionId: 1 });
maintenanceTaskSchema.index({ criticality: 1 });
maintenanceTaskSchema.index({ dueDate: 1 });

module.exports = mongoose.model('MaintenanceTask', maintenanceTaskSchema);

File: backend/models/maintenance_task.py
"""
MaintenanceTask model — replaces Mongoose MaintenanceTask.js.
Extended with real ingestion fields, 5-feature AI scoring, and inspection gap tracking.
"""
import uuid
from datetime import date, datetime
from sqlalchemy import (
    String, Boolean, Integer, Float, Date, DateTime, Text,
    ForeignKey, Enum, Index, func, Uuid, JSON,
)
from sqlalchemy.orm import Mapped, mapped_column
from backend.db.base import Base


class MaintenanceTask(Base):
    __tablename__ = "maintenance_tasks"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    task_id: Mapped[str] = mapped_column(String(60), unique=True, nullable=False, index=True)

    # Source system
    source_system: Mapped[str] = mapped_column(
        Enum("TMS", "SMMS", "TDMS", name="source_system_enum", native_enum=False), nullable=False
    )
    department: Mapped[str] = mapped_column(
        Enum("Engineering", "Traction Distribution", "Signal & Telecom", name="dept_enum", native_enum=False),
        nullable=False,
    )
    section_id: Mapped[str] = mapped_column(
        String(50), ForeignKey("corridor_blocks.section_id"), nullable=False
    )
    section_name: Mapped[str] = mapped_column(String(200), default="")
    zone_code: Mapped[str | None] = mapped_column(String(20), default="NR", index=True)

    # Defect details
    defect_type: Mapped[str] = mapped_column(String(200), nullable=False)
    defect_description: Mapped[str] = mapped_column(Text, default="")
    criticality: Mapped[str] = mapped_column(
        Enum("critical", "high", "medium", "low", name="criticality_enum", native_enum=False),
        default="medium",
    )
    reported_date: Mapped[date] = mapped_column(Date, nullable=False)
    due_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    estimated_duration: Mapped[int] = mapped_column(Integer, nullable=False)  # minutes
    location_km: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Status
    status: Mapped[str] = mapped_column(
        Enum(
            "pending", "scheduled", "approved", "in_progress", "completed", "cancelled",
            name="task_status_enum",
            native_enum=False,
        ),
        default="pending",
        index=True,
    )

    # AI scoring (5-feature model)
    criticality_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    urgency_tier: Mapped[str | None] = mapped_column(String(20), nullable=True)
    score_breakdown: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # {"safety": 0.9, "overdue": 0.3, "traffic": 1.0, "recurrence": 0.4, "inspection_gap": 0.6}
    ai_reasoning: Mapped[str | None] = mapped_column(Text, nullable=True)
    ml_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    rule_score: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Historical recurrence
    recurrence_count: Mapped[int] = mapped_column(Integer, default=0)
    last_occurrence: Mapped[date | None] = mapped_column(Date, nullable=True)
    inspection_gap_days: Mapped[int] = mapped_column(Integer, default=0)

    # Data ingestion
    import_source: Mapped[str] = mapped_column(
        Enum("manual", "csv", "api", "seed", "data.gov.in", name="import_source_enum", native_enum=False), default="manual"
    )
    import_batch_id: Mapped[str | None] = mapped_column(String(60), nullable=True)

    # Assignment
    assigned_to: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("users.id"), nullable=True
    )
    notes: Mapped[str] = mapped_column(Text, default="")

    created_at: Mapped[datetime] = mapped_column(DateTime, default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=func.now(), onupdate=func.now()
    )

    __table_args__ = (
        Index("ix_tasks_dept_status", "department", "status"),
        Index("ix_tasks_section", "section_id"),
        Index("ix_tasks_criticality", "criticality"),
    )

    def __repr__(self) -> str:
        return f"<Task {self.task_id} [{self.criticality}] {self.defect_type}>"
